<a href="https://colab.research.google.com/github/NayraSousa/mrfi-teste/blob/dev/lenet.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [79]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [1]:
!pip install mrfi
!pip install torch
!pip install torchvision

In [2]:
from mrfi import MRFI, EasyConfig
import torch
import torchvision
from torchvision import transforms
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from mrfi.experiment import Acc_experiment, Acc_golden
import math
import random
import csv

device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

In [3]:
class LeNet(nn.Module):
  def __init__(self, trained=False):
    super(LeNet, self).__init__()
    self.conv1 = nn.Conv2d(1, 6, 5)
    self.conv2 = nn.Conv2d(6, 16, 5)
    self.fc1 = nn.Linear(16*4*4, 120)
    self.fc2 = nn.Linear(120, 84)
    self.fc3 = nn.Linear(84, 10)

    if trained:
      self.load_state_dict(torch.load('/content/drive/MyDrive/LeNet/train/lenet_mnist.pth'))

  def forward(self, x):
    x = F.max_pool2d(F.relu(self.conv1(x)), (2,2))
    x = F.max_pool2d(F.relu(self.conv2(x)), 2)
    x = x.view(x.size()[0], -1)
    x = F.relu(self.fc1(x))
    x = F.relu(self.fc2(x))
    x = self.fc3(x)
    return x

def train(epochs=3):
  lenet.to(device)

  criterion = nn.CrossEntropyLoss()
  optimizer = optim.Adam(lenet.parameters(), lr=0.001)

  lenet.train()

  for epoch in range(epochs):
    running_loss = 0
    batch_size = 100
    for i, data in trainloader:
      inputs, labels = i.to(device), data.to(device)

      optimizer.zero_grad()

      outputs = lenet(inputs)
      loss = criterion(outputs, labels)
      loss.backward()
      optimizer.step()
      running_loss += loss.item()

    print(f"Epoch {epoch+1}/{epochs}, Loss: {running_loss/len(trainloader):.4f}")
  torch.save(lenet.state_dict(), '/content/drive/MyDrive/LeNet/train/lenet_mnist.pth')

def test():
  lenet = LeNet().to(device)
  lenet.load_state_dict(torch.load('/content/drive/MyDrive/LeNet/train/lenet_mnist.pth'))
  lenet.eval()

  correct = 0
  total = 0

  with torch.no_grad():
      for images, labels in testloader:
          images, labels = images.to(device), labels.to(device)
          outputs = lenet(images)
          _, predicted = torch.max(outputs.data, 1)
          total += labels.size(0)
          correct += (predicted == labels).sum().item()
          break

      print('Accuracy of the network on the 10000 test images: %d %%' % (
        100 * correct / total))

def get_mnist(batch_size=64):
  transform = transforms.Compose(
      [transforms.ToTensor(),
       transforms.Normalize((0.1307,), (0.3081,))]
  )

  trainset = torchvision.datasets.MNIST(
      root='/content/drive/MyDrive/LeNet/data', train=True, download=True, transform=transform
  )
  testset = torchvision.datasets.MNIST(
      root='/content/drive/MyDrive/LeNet/data', train=False, download=True, transform=transform
  )

  trainloader = torch.utils.data.DataLoader(
      trainset, batch_size=batch_size, shuffle=True
  )
  testloader = torch.utils.data.DataLoader(
      testset, batch_size=batch_size, shuffle=False
  )

  return trainloader, testloader, testset

def load_dataset(dataset='mnist', batch_size=64):

  if dataset.lower()=='mnist':
    train, test, testset = get_mnist()

    return train, test, testset

In [4]:
def calculates_number_positions(e, N, t, p=0.5):

  denominador = 1 + (e ** 2) * ((N-1) / ((t**2*p*(1-p))))
  N_inj = N / denominador

  return N_inj

def return_abs_value(model, layer_name, position):
  layer = getattr(model, layer_name)
  flattened_weights = layer.weight.data.flatten()
  magnitude = torch.abs(flattened_weights[position]).item()

  return magnitude

def return_fi_model(model, econfig):
  fi_model = MRFI(lenet.eval(), econfig)
  fi_model.to(device)
  return fi_model

def return_calculation_times(model, layer_name, input_size=28):
  if layer_name == 'conv1':
    OFW = ((input_size + 2*lenet.conv1.padding[0] - lenet.conv1.kernel_size[0]) // lenet.conv1.stride[0]) + 1
    OFH = OFW
    CT_i = OFW * OFH

    return CT_i

  if layer_name == 'fc3':
    return 1

In [5]:
def evaluate_with_injection(fi_model, loader):
    fi_model.eval()
    correct_gold = 0
    correct_inj  = 0
    total = 0

    fi_model.observers_reset()

    for images, labels in loader:
        images = images.to(device)
        labels = labels.to(device)
        total += labels.size(0)

        with fi_model.golden_run():
            out_golden = fi_model(images)
            _, pred_g = out_golden.max(dim=1)
            correct_gold += (pred_g == labels).sum().item()

        out_inject = fi_model(images)
        _, pred_f = out_inject.max(dim=1)
        correct_inj += (pred_f == labels).sum().item()

    acc_golden = correct_gold / total
    acc_inject = correct_inj  / total

    obs_res = fi_model.observers_result()

    return acc_golden, acc_inject

In [6]:
trainloader, testloader,testset = load_dataset()

lenet = LeNet(trained=True)

for name, param in lenet.named_parameters():
    print(f"Camada: {name}")
    print(f"Parâmetros: {param.count_nonzero()}")
    print("-" * 50)

Camada: conv1.weight
Parâmetros: 150
--------------------------------------------------
Camada: conv1.bias
Parâmetros: 6
--------------------------------------------------
Camada: conv2.weight
Parâmetros: 2400
--------------------------------------------------
Camada: conv2.bias
Parâmetros: 16
--------------------------------------------------
Camada: fc1.weight
Parâmetros: 30720
--------------------------------------------------
Camada: fc1.bias
Parâmetros: 120
--------------------------------------------------
Camada: fc2.weight
Parâmetros: 10080
--------------------------------------------------
Camada: fc2.bias
Parâmetros: 84
--------------------------------------------------
Camada: fc3.weight
Parâmetros: 840
--------------------------------------------------
Camada: fc3.bias
Parâmetros: 10
--------------------------------------------------


In [7]:
n_conv1 = calculates_number_positions(0.05, 150, 2.60)
n_fc3 = calculates_number_positions(0.05, 840, 2.60)

pos_conv1 = random.sample(range(150), int(n_conv1))
print(pos_conv1)

pos_fc3 = random.sample(range(840), int(n_fc3))
print(pos_fc3)

[93, 81, 65, 104, 96, 64, 106, 71, 67, 9, 27, 136, 144, 25, 49, 92, 130, 7, 139, 38, 51, 45, 50, 13, 134, 75, 40, 73, 83, 76, 82, 30, 86, 119, 117, 16, 145, 149, 103, 41, 115, 116, 35, 21, 110, 132, 68, 33, 26, 59, 69, 131, 37, 105, 125, 94, 72, 6, 22, 135, 55, 0, 60, 90, 137, 124, 127, 39, 77, 56, 43, 32, 109, 126, 123, 146, 147, 53, 91, 46, 15, 84, 24, 78, 36, 34, 4, 12, 101, 14, 87, 57, 98, 97, 100, 10, 118, 62, 102, 120, 20, 31, 85, 17, 80, 89, 47, 8, 95, 11, 107, 52, 142, 138, 23, 111, 79, 99, 143, 48, 108, 113]
[37, 570, 261, 432, 478, 584, 422, 516, 389, 74, 338, 61, 21, 310, 247, 289, 84, 812, 327, 336, 718, 482, 654, 175, 755, 58, 148, 534, 722, 378, 81, 379, 355, 596, 311, 303, 290, 490, 782, 428, 555, 554, 348, 825, 4, 551, 77, 186, 634, 830, 334, 642, 829, 23, 586, 509, 569, 713, 230, 293, 283, 454, 681, 127, 521, 416, 318, 214, 631, 241, 153, 285, 304, 301, 595, 513, 122, 694, 547, 571, 52, 364, 139, 444, 257, 475, 460, 492, 592, 54, 38, 220, 392, 665, 715, 225, 732, 488, 

In [8]:
bits = [6, 7]
layers = ['conv1', 'fc3']
metadata = []

CT_conv1 = return_calculation_times(lenet, layers[0])
CT_fc3 = return_calculation_times(lenet, layers[1])

In [ ]:
for n in bits:
  for name in layers:

    if name == 'conv1':
      for position in pos_conv1:

        lenet = LeNet(trained=True)
        magnitude = return_abs_value(lenet, 'conv1', position)

        config_str = f"""
                    faultinject:
                      - type: weights
                        name: [weight]
                        quantization:
                          method: SymmericQuantization
                          bit_width: 8
                          dynamic_range: auto
                        selector:
                          method: FixPosition
                          position: {position}
                        error_mode:
                          method: IntFixedBitFlip
                          bit_width: 8
                          bit: {n}
                        module_name: conv1
                    """

        econfig = EasyConfig.load_string(config_str)
        fi_model = return_fi_model(lenet, econfig)
        acc_g, acc_f = evaluate_with_injection(fi_model, testloader)

        metadata.append({f'Camada': name, 'Bit': econfig.faultinject[0]['error_mode']['args']['bit'],
                         'posicao': econfig.faultinject[0]['selector']['args']['position'], 'magnitude': magnitude,
                         'vulnerabilidade': acc_g-acc_f, 'calculation_times': CT_conv1})

    if name == 'fc3':
      for position in pos_fc3:

        lenet = LeNet(trained=True)
        magnitude = return_abs_value(lenet, 'fc3', position)

        config_str = f"""
                    faultinject:
                        - type: weights
                          name: [weight]
                          quantization:
                            method: SymmericQuantization
                            bit_width: 8
                            dynamic_range: auto
                          selector:
                            method: FixPosition
                            position: {position}
                          error_mode:
                            method: IntFixedBitFlip
                            bit_width: 8
                            bit: {n}
                          module_name: fc3
                          """

        econfig = EasyConfig.load_string(config_str)
        fi_model = return_fi_model(lenet, econfig)
        acc_g, acc_f = evaluate_with_injection(fi_model, testloader)

        metadata.append({f'Camada': name, 'Bit': econfig.faultinject[0]['error_mode']['args']['bit'],
                         'posicao': econfig.faultinject[0]['selector']['args']['position'], 'magnitude': magnitude,
                         'vulnerabilidade': acc_g-acc_f, 'calculation_times': CT_fc3})

with open('/content/drive/MyDrive/LeNet/teste.csv', 'w', newline='', encoding='utf-8') as file:
    escritor = csv.DictWriter(file, fieldnames=metadata[0].keys())
    escritor.writeheader()
    escritor.writerows(metadata)

In [51]:
lenet = LeNet(trained=True)
config_str = f"""
                    faultinject:
                        - type: weights
                          name: [weight]
                          quantization:
                            method: SymmericQuantization
                            bit_width: 8
                            dynamic_range: auto
                          selector:
                            method: FixPositions
                            positions: [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106, 111, 133, 123, 118, 112, 20, 127, 40, 128, 117, 7, 116]
                          error_mode:
                            method: IntFixedBitFlip
                            bit_width: 8
                            bit: 7
                          module_name: conv1

                        - type: weights
                          name: [weight]
                          quantization:
                            method: SymmericQuantization
                            bit_width: 8
                            dynamic_range: auto
                          selector:
                            method: FixPositions
                            positions: [812, 826, 818, 788, 784, 796, 816, 444, 596, 74, 776, 653, 829, 493, 595, 438, 498, 420, 660, 503, 483, 448, 442, 467, 227]
                          error_mode:
                            method: IntFixedBitFlip
                            bit_width: 8
                            bit: 7
                          module_name: fc3
                          """
econfig = EasyConfig.load_string(config_str)
fi_model = MRFI(lenet.eval(), econfig)
fi_model.to(device)

acc_g, acc_f = evaluate_with_injection(fi_model, testloader)
acc_g, acc_f

(0.9863, 0.2673)

# CONV1

0.9795 [11]

0.9838 [11, 103]

0.9797 [11, 103, 5]

0.9852 [11, 103, 5, 108]

0.9754 [11, 103, 5, 108, 1]

0.9476 [11, 103, 5, 108, 1, 104]

0.973 [11, 103, 5, 108, 1, 104, 107]

0.9084 [11, 103, 5, 108, 1, 104, 107, 10]

0.8319 [11, 103, 5, 108, 1, 104, 107, 10, 6]

0.892 [11, 103, 5, 108, 1, 104, 107, 10, 6, 21]

## Tirando os gradientes negativos em ordem decrescente:

0.8272 [11, 103, 5, 1, 104, 107, 10, 6, 21]

0.7689 [11, 103, 5, 1, 104, 10, 6, 21]

0.8499 [11, 103, 5, 1, 10, 6, 21]

0.9125 [11, 5, 1, 10, 6, 21]

## Botando em ordem crescente de gradient:

0.9125 [5, 10, 1, 6, 21, 11]

## Botando em ordem decrescente de vulnerabilidade:

0.9125 [11, 5, 1, 10, 6, 21]

## Menor acurácia:

Foi dado pelas seguintes posições: [11, 103, 5, 1, 104, 10, 6, 21]. Agora vou testar com essas posições e o restante das 15 mais vulneravéis.

## Adicionando as posições a partir da 11° mais vulnerável:

0.7023 [11, 103, 5, 1, 104, 10, 6, 21, 124]

0.7773 [11, 103, 5, 1, 104, 10, 6, 21, 124, 109]

0.7444  [11, 103, 5, 1, 104, 10, 6, 21, 124, 109, 0]

0.6529 [11, 103, 5, 1, 104, 10, 6, 21, 124, 109, 0, 119]

0.6246 [11, 103, 5, 1, 104, 10, 6, 21, 124, 109, 0, 119, 15]

## Tirando os gradientes negativos em ordem decrescente:

0.5582 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15]

0.6446 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 15]

0.7108  [11, 103, 5, 1, 104, 10, 6, 21, 0, 15]

## Botando em ordem crescente de gradient:

0.8598 [15, 5, 10, 0, 1, 6, 21, 11]

## Botando em ordem crescente de vulnerabilidade:

0.8598 [11, 5, 1, 10, 6, 21, 0, 15]

## Testando a partir do 16° bit mais vulneravel:

0.5333 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106, 111, 114, 133]

0.5441  [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106, 111, 114]

0.501 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106, 111]

0.5411 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106]

0.5582 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15]

## Tirando os gradientes negativos em ordem decrescente:

0.4951 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106, 111, 133]

0.501  [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106, 111]

0.5411 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15, 106]

0.5582 [11, 103, 5, 1, 104, 10, 6, 21, 124, 0, 119, 15]